File: backend/app/schemas/project.py
from __future__ import annotations
from datetime import datetime
from pydantic import Field
from app.models.enums import ProjectAccessStatusEnum, ProjectMemberRoleEnum
from .common import BaseSchema, IdSchema, TimestampSchema
from .user import UserShortRead


class ProjectAccessRequestCreate(BaseSchema):
    member_role: ProjectMemberRoleEnum = ProjectMemberRoleEnum.VIEWER
    request_note: str | None = Field(None, max_length=1000)


class ProjectAccessReviewRequest(BaseSchema):
    member_role: ProjectMemberRoleEnum | None = None
    review_note: str | None = Field(None, max_length=1000)


class ProjectMemberCreate(BaseSchema):
    user_id: int
    member_role: ProjectMemberRoleEnum
    request_note: str | None = Field(None, max_length=1000)


class ProjectMemberUpdate(BaseSchema):
    member_role: ProjectMemberRoleEnum | None = None
    review_note: str | None = Field(None, max_length=1000)


class ProjectMemberRead(IdSchema):
    project_id: int
    user_id: int
    member_role: ProjectMemberRoleEnum
    access_status: ProjectAccessStatusEnum
    added_by: int | None = None
    added_at: datetime
    requested_at: datetime
    request_note: str | None = None
    reviewed_by: int | None = None
    reviewed_at: datetime | None = None
    review_note: str | None = None


class ProjectMemberDetailRead(ProjectMemberRead):
    user: UserShortRead
    creator: UserShortRead | None = None
    reviewer: UserShortRead | None = None


class ProjectCreate(BaseSchema):
    name: str = Field(..., min_length=2, max_length=255)
    code: str = Field(..., min_length=2, max_length=100)
    description: str | None = None
    owner_id: int


class ProjectUpdate(BaseSchema):
    name: str | None = Field(None, min_length=2, max_length=255)
    code: str | None = Field(None, min_length=2, max_length=100)
    description: str | None = None
    is_archived: bool | None = None


class ProjectRead(TimestampSchema, IdSchema):
    name: str
    code: str
    description: str | None = None
    owner_id: int
    is_archived: bool


class ProjectDetailRead(ProjectRead):
    owner: UserShortRead
    members: list[ProjectMemberDetailRead] = []

File: frontend/src/shared/types/project.ts
export type ProjectMemberRole = 'owner' | 'manager' | 'editor' | 'viewer';
export type ProjectAccessStatus = 'requested' | 'approved' | 'rejected';

export interface ProjectUserSummary {
  id: number;
  email: string;
  full_name: string;
  position: string | null;
  department: string | null;
  is_active: boolean;
}

export type ProjectAvailableUser = ProjectUserSummary;

export interface Project {
  id: number;
  name: string;
  code: string;
  description: string | null;
  owner_id: number;
  is_archived: boolean;
  created_at: string;
  updated_at: string;
}

export interface ProjectDetail extends Project {
  owner: ProjectUserSummary | null;
}

export interface ProjectMember {
  id: number;
  project_id: number;
  user_id: number;
  member_role: ProjectMemberRole;
  access_status: ProjectAccessStatus;
  added_by: number | null;
  added_at: string;
  requested_at: string;
  request_note: string | null;
  reviewed_by: number | null;
  reviewed_at: string | null;
  review_note: string | null;
  user: ProjectUserSummary;
  creator?: ProjectUserSummary | null;
  reviewer?: ProjectUserSummary | null;
}

export interface CreateProjectPayload {
  name: string;
  code: string;
  description?: string | null;
  owner_id: number;
}

export interface AddProjectMemberPayload {
  user_id: number;
  member_role: ProjectMemberRole;
  request_note?: string | null;
}

export interface UpdateProjectMemberPayload {
  member_role?: ProjectMemberRole;
  review_note?: string | null;
}

export interface RequestProjectAccessPayload {
  member_role: ProjectMemberRole;
  request_note?: string | null;
}

export const PROJECT_MEMBER_ROLE_OPTIONS: Array<{
  value: ProjectMemberRole;
  label: string;
}> = [
  { value: 'viewer', label: 'Наблюдатель' },
  { value: 'editor', label: 'Редактор' },
  { value: 'manager', label: 'Менеджер' },
];

export function getProjectRoleLabel(role: ProjectMemberRole) {
  switch (role) {
    case 'owner':
      return 'Владелец';
    case 'manager':
      return 'Менеджер';
    case 'editor':
      return 'Редактор';
    case 'viewer':
      return 'Наблюдатель';
    default:
      return role;
  }
}

export function getProjectAccessStatusLabel(status: ProjectAccessStatus) {
  switch (status) {
    case 'requested':
      return 'Запрошен';
    case 'approved':
      return 'Одобрен';
    case 'rejected':
      return 'Отклонен';
    default:
      return status;
  }
}